processing/src/main/java/io/druid/query/groupby/epinephelinae/GroupByMergingQueryRunnerV2.java

```diff
@@ -0,0 +1,661 @@
+
+                          // Add dimensions.
+                          for (int i = 0; i < entry.getKey().getDimensions().length; i++) {
+                            theMap.put(
+                                query.getDimensions().get(i).getOutputName(),
+                                Strings.emptyToNull(entry.getKey().getDimensions()[i])
+                            );
+                          }
+
+                          // Add aggregations.
+                          for (int i = 0; i < entry.getValues().length; i++) {
+                            theMap.put(query.getAggregatorSpecs().get(i).getName(), entry.getValues()[i]);
+                          }
+
+                          return new MapBasedRow(
+                              query.getGranularity().toDateTime(entry.getKey().getTimestamp()),
+                              theMap
+                          );
+                        }
+                      }
+                  );
+                }
+
+                @Override
+                public void cleanup(CloseableGrouperIterator<GroupByMergingKey, Row> iterFromMake)
+                {
+                  iterFromMake.close();
+                }
+              }
+          ),
+          new Closeable()
+          {
+            @Override
+            public void close() throws IOException
+            {
+              mergeBufferHolder.close();
+              CloseQuietly.close(temporaryStorage);
+            }
+          }
+      );
+    } catch (Exception e) {
+      // Exception caught while creating the sequence; release mergeBufferHolder.
+      mergeBufferHolder.close();
+      throw e;
+    }
+  }
+
+  private void waitForFutureCompletion(
+      GroupByQuery query,
+      ListenableFuture<?> future,
+      long timeout
+  )
+  {
+    try {
+      if (queryWatcher != null) {
+        queryWatcher.registerQuery(query, future);
+      }
+      if (timeout <= 0) {
+        future.get();
+      } else {
+        future.get(timeout, TimeUnit.MILLISECONDS);
+      }
+    }
+    catch (InterruptedException e) {
+      log.warn(e, "Query interrupted, cancelling pending results, query id [%s]", query.getId());
+      future.cancel(true);
+      throw new QueryInterruptedException(e);
+    }
+    catch (CancellationException e) {
+      throw new QueryInterruptedException(e);
+    }
+    catch (TimeoutException e) {
+      log.info("Query timeout, cancelling pending results for query id [%s]", query.getId());
+      future.cancel(true);
+      throw new QueryInterruptedException(e);
+    }
+    catch (ExecutionException e) {
+      throw Throwables.propagate(e.getCause());
+    }
+  }
+
+  private static class GroupByMergingKey implements Comparable<GroupByMergingKey>
+  {
+    private final long timestamp;
+    private final String[] dimensions;
+
+    @JsonCreator
+    public GroupByMergingKey(
+        @JsonProperty("t") long timestamp,
+        @JsonProperty("d") String[] dimensions
```

Comment:
it's just to reduce the serialized size (as there will be a LOT of these). will add a comment.